include/adc.h

```diff
@@ -33,8 +33,16 @@ extern "C" {
  * to define a sample from a specific channel.
  */
 struct adc_seq_entry {
+#if defined(CONFIG_ADC_LEGACY_API)
 	/** Clock ticks delay before sampling the ADC. */
 	s32_t sampling_delay;
+#else
+	/** Delay before starting sampling the ADC in microseconds. */
+	u32_t sampling_delay;
```

Comment:
Why do we need sampling_delay? Can't the user simply call adc_read() later?

Reply:
@mnkp You are right it is simpler but only for single adc_seq_entry, if you have multiple sequence entries I guess it is more convenient to specify delay between entries and then just call adc_read() once.

Reply:
Isn't it what 'sampling_interval' is for?

Reply:
@mnkp Each adc_seq_entry has its own sampling interval. For example you take 5 samples of channel 1 with sampling interval 40 ms, then you want to wait 100 ms and take 10 samples of channel 2 with sampling interval 20 ms.

Reply:
The behavior you describe is not clear when looking at the description of the sampling_delay, sampling_interval fields. In any case, I think we are going the wrong direction with this change.

A very typical case when using a multi-channel ADC is to sample several channels simultaneously (or in fact relatively close to each other, with the delay required for the conversion time) then wait some time and sample again. In this case sampling rate is the frequency with which each of this multi-channel measurement is triggered. One example would be sampling of the stereo audio signal but anything else be it series of analog pressure sensors, potentiometers would be the same. We almost never want to sample one channel several times then sample the next (but we surely should support this case too).

Reply:
@mnkp I absolutely agree with you. This is just another example of poor driver API. Besides no support for the situation you have described there is also no way for dynamic setting of resolution. 

I just wanted to add small improvement not redesign it. I get the impression that no one cares about drivers APIs and drivers codes. Right now I feel there is only one complete API (SPI), look how long ago it was updated and still no one works on updating the drivers. I have been working on 4 drivers: SPI, RTCounter, ADC, Watchdog. In 3 of 4 cases I got the feeling that they were just copy/pasted from some specific platform where they just met limited expectations.

Reply:
@m-kru I have to agree with you. While I'm fond of Zephyr, about any aspect of it, its driver APIs are certainly substandard: poorly documented, inconsistent, missing features and not very good in supporting typical use cases. ADC API is a good example.

That said, designing a good API is not easy at all and I'm afraid ADC API will be especially tricky. While most of protocol APIs like I2C, I2S, SPI are well defined and the hardware blocks which implement them are self contained that's not a case with ADC.

Let's talk about implementing ADC timings. A typical multi-channel ADC hardware module operates as follows:
- configure channels on which measurements need to be taken
- wait for a hardware or software trigger
- take measurements on all preconfigured channels with some delay required for the conversion of each channel

From the point of view of ADC module sampling rate corresponds to the time required for the conversion of each channel because that's how often each sample is taken. However, from the point of view of a user of ADC API sampling rate is the frequency of the hardware of software trigger. If we want to sample stereo audio signal at 8kHz or take measurement of 4 analog pressure sensors at 10 Hz that's our sampling frequency. We don't care much what's the conversion time required to measure one single channel.

As I mentioned ADC modules often implement a hardware trigger which can start a (possibly multi-channel) measurement. This trigger will typically be a GPIO pin or a timer. Timers are external to ADC module, often only predefined few can be used to trigger ADC conversion, e.g. out of 4 timers present in the SoC only timer0 or timer1. This creates external dependencies and complicates the API. We can't simply define an arbitrary sampling interval and expect that the right thing will happen. That's not implementable.

Conversion time is strictly dependent on ADC module clock and typically can vary only in a very limited range. E.g. in case of Intel Quark microcontroller on which current ADC API is based it can vary between 10 and 266 ADC clocks for 8-bit resolution. That's very typical, other ADC modules are the same. So also here we should not expect to be able to set arbitrary conversion time rates.

As for the 'conversion time' maybe that not the best name. The ADC terminology is a bit complicated. Conversion time often refers strictly to the time required to convert analog signal to a digital value but there is also tracking time which is the length of time when ADC input is connected to any given channel, etc. Maybe we should talk about single channel sampling time and correspondingly about single channel sampling rate.

Reply:
@mnkp ADC API just needs more thorough update. I am withdrawing this PR.